### DISCUSSION TITLE
After gc.collect, mem_free is more than 1Mbytes but cannot allocate 200kb bytearray.

### DISCUSSION THREAD

**Discussion by sam0910** [micropython/micropython#12987] [ESP32] [answered]
Greetings,

After using graphic lib(lvgl), load some png etc..
unload the graphics and free some memory with,
gc.collect(), gc.mem_free() is more than 1Mbytes but I cannot allocate 200kb bytearray.

I am so lost, please show me the light :)

```
>>> gc.collect()
>>> gc.mem_alloc()
510288
>>> gc.mem_free()
1538640
>>> cbuf = bytearray(230400) 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
MemoryError: memory allocation failed, allocating 230400 bytes
```

**Comment by andrewleech** [comment, marked as answer]
This sort of issue is quite common in longer-running applications, or ones that have a lot of "object churn" - it's memory fragmentation.

It's an issue on pretty much all embedded systems of any kind, unless they implement a (dramatically more complicated) dynamic memory manager that can perform deframenting.

There were quite a lot of discussions about this on the old forum: https://www.google.com/search?q=micropython+memory+fragmentation+site:forum.micropython.org

**Reply by sam0910** [reply]
Thank you so much, I'll follow up the path :)
